homeassistant/components/ovo_energy/__init__.py

```diff
@@ -0,0 +1,112 @@
+"""Support for OVO Energy."""
+import logging
+from typing import Any, Dict
+
+import aiohttp
+from ovoenergy.ovoenergy import OVOEnergy
+
+from homeassistant.components.ovo_energy.const import (
+    CONF_ACCOUNT_ID,
+    DATA_OVO_ENERGY_CLIENT,
+    DOMAIN,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_PASSWORD, CONF_USERNAME
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.typing import ConfigType, HomeAssistantType
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
+    """Set up the OVO Energy components."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistantType, entry: ConfigEntry) -> bool:
+    """Set up OVO Energy from a config entry."""
+
+    client = OVOEnergy()
+
+    try:
+        await client.authenticate(entry.data[CONF_USERNAME], entry.data[CONF_PASSWORD])
+    except aiohttp.ClientError as exception:
+        _LOGGER.warning(exception)
+        raise ConfigEntryNotReady from exception
+
+    instance_key = f"{DOMAIN}_{entry.data[CONF_ACCOUNT_ID]}"
+    hass.data.setdefault(instance_key, {})[DATA_OVO_ENERGY_CLIENT] = client
+
+    # Setup components
+    hass.async_create_task(
+        hass.config_entries.async_forward_entry_setup(entry, "sensor")
+    )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistantType, entry: ConfigType) -> bool:
+    """Unload OVO Energy config entry."""
+    # Unload sensors
+    await hass.config_entries.async_forward_entry_unload(entry, "sensor")
+
+    del hass.data[f"{DOMAIN}_{entry.data[CONF_ACCOUNT_ID]}"]
+
+    return True
+
+
+class OVOEnergyEntity(Entity):
+    """Defines a base OVO Energy entity."""
+
+    def __init__(self, client: OVOEnergy, key: str, name: str, icon: str) -> None:
+        """Initialize the OVO Energy entity."""
+        self._client = client
+        self._key = key
+        self._name = name
+        self._icon = icon
+        self._available = True
+
+    @property
+    def name(self) -> str:
+        """Return the name of the entity."""
+        return self._name
+
+    @property
+    def icon(self) -> str:
+        """Return the mdi icon of the entity."""
+        return self._icon
+
+    @property
+    def available(self) -> bool:
+        """Return True if entity is available."""
+        return self._available
```

Comment:
If you pass in the coordinator into the base class, you can simply check `self.coordinator.last_update_success` to use as the availability state.

Reply:
I've joined the two as the sensors do some additional checks if the data exists. Also done some cleanup around this